### ISSUE TITLE
retrain speculator

### ISSUE THREAD

**Issue by changhoonhahn** [changhoonhahn/gqp_mc#52]
The current version of the speculator FSPS emulator is inaccurate in the SFH basis coefficient parameter space that we're interested in (see [speculator_accuracy.ipynb](https://github.com/changhoonhahn/gqp_mc/blob/7e66c781395d6bf3095a172cb0af876406c9d819/nb/speculator_accuracy.ipynb)). We need to **update the speculator FSPS emulator and validate its accuracy.**

**Comment by changhoonhahn** [comment]
The `speculator` package now has a demo for training speculator: [speculator_training_demo.ipynb](https://github.com/justinalsing/speculator/blob/master/speculator_training_demo.ipynb). 
- [ ] Retrain emulator following its instructions. 
- [ ] Validate retrained emulator

**Comment by changhoonhahn** [comment]
Scripts for generating `speculator` training set for simple (calzetti) dust model has been implemented and deployed on NERSC (d7bd33ea94c987f61690582579c767a274284bc7): 
- [speculator_training.py](https://github.com/changhoonhahn/gqp_mc/blob/f81f592cb7bbc4198e236406d885bb3b90d6ce8f/run/speculator_training.py)
- [deploy_speculator_trainingset.py](https://github.com/changhoonhahn/gqp_mc/blob/f81f592cb7bbc4198e236406d885bb3b90d6ce8f/run/cori/deploy_speculator_trainingset.py)

**Comment by changhoonhahn** [comment]
@kgb0255 lets try to keep track of the speculator training using the task list I've added here: https://github.com/changhoonhahn/gqp_mc/blob/master/run/todo.speculator_training.md

As you train speculator models with different training set sizes, pca components, etc, lets fill out the table to keep track.

**Comment by changhoonhahn** [comment]
**simpledust, 300 training batches, 20 PCAs, 200 pca training batches: ~4% accuracy for 99 percentile**
![image](https://user-images.githubusercontent.com/1895640/96885856-01cc3a00-1451-11eb-801a-9ad7f22bf5d2.png)

**Comment by kgb0255** [comment]
**simpledust, 300 training batches, 20 PCAs, 200 pca training batches: ~4% accuracy for 99 percentile**
![image](https://user-images.githubusercontent.com/57269223/96972730-bb67f100-14cb-11eb-8188-cd097134a685.png)

**simpledust, 300 training batches, 40 PCAs, 200 pca training batches: ~5% accuracy for 99 percentile**
![james_validate_desi_simpledust 0_299 pca40 1](https://user-images.githubusercontent.com/57269223/96972959-16014d00-14cc-11eb-9be0-c513b296a9aa.png)

**Comment by kgb0255** [comment]
I've uploaded the collection of dPCA plots [here](https://github.com/kgb0255/GQPMC_v4_JAMES/tree/main/dPCA_plots).

**Comment by changhoonhahn** [comment]
# PCAs in intervals
Since we've been struggling to train speculator for the entire wavelength range (2305 - 11025), let's try to divide it into three wavelength ranges and train a NN separately for each one. 

Here are the wavelength ranges I've settled on: 
1. 2305 < wave < 4500
2. 4500 < wave < 6500
3. 6500 < wave < 11025
These were chosen based on the accuracy level at different wavelengths (_e.g._ we struggle to get good accuracy at < 4500) and so that each bin has similar dimensions. 

I've added two notebooks to train and validate speculator in separate wavelength bins: 
- [training_desi_complexdust_speculator_wavebins.ipynb](https://github.com/changhoonhahn/gqp_mc/blob/b83dcfb0d2e17bfcebeed7d53250019138b135a4/nb/training_desi_complexdust_speculator_wavebins.ipynb): for training the speculator model for a specific wavelength bin
- [validate_trained_desi_complexdust_speculator_wavebins.ipynb](https://github.com/changhoonhahn/gqp_mc/blob/b83dcfb0d2e17bfcebeed7d53250019138b135a4/nb/validate_trained_desi_complexdust_speculator_wavebins.ipynb): for validating the separate wavelength bin speculator model

**Comment by changhoonhahn** [comment]
Not sure why I only included wave > 3000 in the validation plot in the past, but we definitely speculator to reach ~1% accuracy over the entire 2305 - 11025 wavelength. The lower and higher wavelengths, make things worse...
![image](https://user-images.githubusercontent.com/1895640/98832418-cd71eb00-240a-11eb-91c9-b5411d11db27.png)

**Comment by changhoonhahn** [comment]
Turns out the situation is not so dire. Here's an updated validation plots with a larger test set (1e5 test samples): 
![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 frac_err](https://user-images.githubusercontent.com/1895640/99119871-ef649c80-25c7-11eb-9670-b9372189013d.png)
**The dip we see above is due to numerical precision issues.**

![image](https://user-images.githubusercontent.com/1895640/99119948-0dca9800-25c8-11eb-874d-f757668c062d.png)
Our mean fractional error isn't at <1% yet, but at 1.5% so we're almost there!

**Comment by changhoonhahn** [comment]
# <1% achieved!
Finally, we've achieved <1% accuracy for wavebin0. Using **Npca=50, Ntrain=5e6, and a [256, 256, 256, 256] architecture** did the trick. 
![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 frac_err](https://user-images.githubusercontent.com/1895640/99264424-6e3b1e80-27ee-11eb-8936-3224255e8b76.png)

![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 mean_frac_err_dist](https://user-images.githubusercontent.com/1895640/99264431-7004e200-27ee-11eb-9f9d-d80b3d9a2ed0.png)
<1% for 99 percentile. 

For consistency, I'm running wavebin1 and wavebin2 with the same architecture.

**Comment by changhoonhahn** [comment]
## Ntrain=5,000,000, Npca=[50, 30, 30], architecture=4x256

![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 frac_err](https://user-images.githubusercontent.com/1895640/99434005-49bd7000-28dc-11eb-9de0-e4c824e60336.png)

![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 mean_frac_err_dist](https://user-images.githubusercontent.com/1895640/99434088-66f23e80-28dc-11eb-9b2b-1d04a8e16047.png)

**Comment by changhoonhahn** [comment]
Ha! I mixed up log base 10 with natural log, so we actually do way better...
![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 frac_err](https://user-images.githubusercontent.com/1895640/99442061-c81f0f80-28e6-11eb-9756-a57015376982.png)
![DESI_complexdust_model Ntrain5000000 wave_bin0 pca50 4x256 mean_frac_err_dist](https://user-images.githubusercontent.com/1895640/99442066-c9e8d300-28e6-11eb-9869-8048bf0d5d59.png)
